CI: Use Mergify for automated PR merging with safety requirements

okurz opened the pull request:
Motivation:
Improve development workflow by automating the merging process once all CI
checks pass. This ensures that PRs are not merged manually with missing or
failed CI checks, which can happen if they are not triggered correctly.
Mergify is more diligent in verifying that all required checks are present and
successful.

Design Choices:
- Require at least 2 approvals for safety, mitigating concerns about bad
  actors or account takeovers.
- Only merge if "integration" and "unit" tests pass on the "master" branch.
- Avoid merging if "not-ready" or "acceptance-tests-needed" labels are present.
- Automated conflict notification to prompt authors for fixes.

User Benefits:
- Faster merging of approved and tested contributions.
- Improved reliability by strictly enforcing CI check requirements.
- Consistent automation across the os-autoinst/openQA organization.

okurz committed:
Add mergify configuration

Martchus commented:
Looks like zypper crashes due to an unhandled C++ exception:

```
terminate called after throwing an instance of 'std::system_error'
  what():  Operation not permitted
```

I suppose this could be considered a zypper bug.

Martchus commented:
Right, I only added it recently to `devel:openQA:Leap:15.3` but that's of course not helping here. We needed a container with our devel repo enabled, maybe just reuse the openQA CI container?

okurz commented:
Rebased, refreshed, updated, verified. All ready for review now.

foursixnine reviewed (changes requested):
Change the approval requirement to members of @os-autoinst/asset-sync-and-integration

okurz commented:
> Change the approval requirement to members of @os-autoinst/asset-sync-and-integration

I don't think that's possible according to https://docs.mergify.com/workflow/actions/merge/ . Also I don't think that we need to overcomplicate things. Still only approvals from priviledged members of the organisation count, not like *anybody*

foursixnine commented:
> > Change the approval requirement to members of @os-autoinst/asset-sync-and-integration
> 
> I don't think that's possible according to https://docs.mergify.com/workflow/actions/merge/ . Also I don't think that we need to overcomplicate things. Still only approvals from priviledged members of the organisation count, not like _anybody_

Not convinced. Ever heard of bad actors and account takeover? - If others (non members of @os-autoinst/tools-team) agree, and number of approvals is increased to 2 you can dismiss.

I am not a fan of automated merges on things that go live - also no description is provided nor reasoning as to _why_ is it important for this repository.

okurz committed:
CI: Use Mergify for automated PR merging with safety requirements

Motivation:
Improve development workflow by automating the merging process once all CI
checks pass. This ensures that PRs are not merged manually with missing or
failed CI checks, which can happen if they are not triggered correctly.
Mergify is more diligent in verifying that all required checks are present and
successful.

Design Choices:
- Require at least 2 approvals for safety, mitigating concerns about bad
  actors or account takeovers.
- Only merge if "integration" and "unit" tests pass on the "master" branch.
- Avoid merging if "not-ready" or "acceptance-tests-needed" labels are present.
- Automated conflict notification to prompt authors for fixes.

User Benefits:
- Faster merging of approved and tested contributions.
- Improved reliability by strictly enforcing CI check requirements.
- Consistent automation across the os-autoinst/openQA organization.

okurz commented:
> > > Change the approval requirement to members of @os-autoinst/asset-sync-and-integration
> > 
> > 
> > I don't think that's possible according to https://docs.mergify.com/workflow/actions/merge/ . Also I don't think that we need to overcomplicate things. Still only approvals from priviledged members of the organisation count, not like _anybody_
> 
> Not convinced. Ever heard of bad actors and account takeover? - If others (non members of @os-autoinst/tools-team) agree, and number of approvals is increased to 2 and I can reconsider.

Sure, let's play it safe(r). Increased to >=2 now. 

> I am not a fan of automated merges on things that go live - also no description is provided nor reasoning as to _why_ is it important for this repository.

Well, first it might seem to be more *convenient* than it is *important* but we have good experiences in os-autoinst, openQA and others and the approach is helping to improve the quality and prevent unforeseen problems because unlike human reviewers mergify is very diligent to ensure that all necessary CI checks are not failed but also *present* as happened multiple times that CI checks were not triggered at all.

foursixnine reviewed (approved):
I still don't think this is needed for this repo, but since we can still merge manually, I have no troubles with the current proposal

Vogtinator commented:
Any news here? FWICT it's still enabled.

okurz commented:
> Any news here? FWICT it's still enabled.

there are no new news. I don't see why this project needs to change anything as we have basically the same rules in other projects for many years. If you have a reasonable proposal considering pros and cons please open an according pull request.

Vogtinator commented:
> I don't see why this project needs to change anything as we have basically the same rules in other projects for many years.

It's a recent change for this (larger) project, and just because it's also done somewhere else does not mean it's good, just that it's widespread.

> If you have a reasonable proposal considering pros and cons please open an according pull request.

That would be a revert of this, with the reasoning being:

> This is highly insecure as long as the GitHub option to automatically remove stale reviews is disabled. See https://github.blog/changelog/2023-06-06-security-enhancements-to-required-approvals-on-pull-requests/
